spyder/plugins/application/plugin.py

```diff
@@ -0,0 +1,310 @@
+from qtpy.QtWidgets import QMenu
+
+# Local imports
+from spyder.api.plugins import Plugins, SpyderPluginV2
+from spyder.api.translations import get_translation
+from spyder.api.widgets.menus import MENU_SEPARATOR
+from spyder.config.base import DEV, get_module_path
+from spyder.plugins.application.confpage import ApplicationConfigPage
+from spyder.plugins.application.container import (
+    ApplicationActions, ApplicationContainer, WinUserEnvDialog)
+from spyder.plugins.mainmenu.api import (
+    ApplicationMenus, FileMenuSections, HelpMenuSections, ToolsMenuSections)
+from spyder.utils.qthelpers import add_actions
+
+# Localization
+_ = get_translation('spyder')
+
+
+class Application(SpyderPluginV2):
+    NAME = 'application'
+    REQUIRES = [Plugins.Console]
+    OPTIONAL = [Plugins.Help, Plugins.MainMenu, Plugins.Preferences,
+                Plugins.Shortcuts]
+    CONTAINER_CLASS = ApplicationContainer
+    CONF_SECTION = 'main'
+    CONF_FILE = False
+    CONF_FROM_OPTIONS = {
+        # Screen resolution section
+        'normal_screen_resolution': ('main', 'normal_screen_resolution'),
+        'high_dpi_scaling': ('main', 'high_dpi_scaling'),
+        'high_dpi_custom_scale_factor': ('main',
+                                         'high_dpi_custom_scale_factor'),
+        'high_dpi_custom_scale_factors': ('main',
+                                          'high_dpi_custom_scale_factors'),
+        # Panes section
+        'vertical_tabs': ('main', 'vertical_tabs'),
+        'use_custom_margin': ('main', 'use_custom_margin'),
+        'custom_margin': ('main', 'custom_margin'),
+        'use_custom_cursor_blinking': ('main', 'use_custom_cursor_blinking'),
+        # Advanced settings
+        'opengl': ('main', 'opengl'),
+        'single_instance': ('main', 'single_instance'),
+        'prompt_on_exit': ('main', 'prompt_on_exit'),
+        'check_updates_on_startup': ('main', 'check_updates_on_startup'),
+        'show_internal_errors': ('main', 'show_internal_errors'),
+    }
+    CONF_WIDGET_CLASS = ApplicationConfigPage
+
+    def get_name(self):
+        return _('Application')
+
+    def get_icon(self):
+        return self.create_icon('genprefs')
+
+    def get_description(self):
+        return _('Provide main application base actions.')
+
+    def register(self):
+        # Register with Preferences plugin
+        preferences = self.get_plugin(Plugins.Preferences)
+        preferences.register_plugin_preferences(self)
+
+        # Application base actions (that interact directly with other
+        # plugins). For other actions see ./container.py
+        self.restart_action = self.create_action(
+            ApplicationActions.SpyderRestart,
+            _("&Restart"),
+            icon=self.create_icon('restart'),
+            tip=_("Restart"),
+            triggered=self.restart,
+            context=Qt.ApplicationShortcut,
+            shortcut_context="_",
+            register_shortcut=True)
```

Comment:
Please move this action to the container and make it emit a signal (`sig_restart_requested`) when triggered. Then connect that signal to the `restart` method here.